qa/tasks/quiescer.py

```diff
@@ -0,0 +1,259 @@
+            if leader is None:
+                self.logger.warn("Couldn't get quiesce db leader from the mds map")
+                self.sleep_unless_stopped(5)
+
+        while leader is not None and not self.is_stopped:
+            command = ['tell', f"mds.{leader}", 'quiesce', 'db']
+            command.extend(args)
+            self.logger.debug("Running ceph command: '%s'" % " ".join(command))
+            result = self.fs.run_ceph_cmd(args=command, check_status=False, stdout=StringIO(), timeoutcmd=self.QDB_CMD_TIMEOUT)
+            rc, stdout = result.exitstatus, result.stdout.getvalue()
+            if rc == -errno.ENOTTY:
+                try:
+                    resp = json.loads(stdout)
+                    leader = int(resp['leader'])
+                    self.logger.info("Retrying a quiesce db command with leader %d" % leader)
+                except Exception as e:
+                    self.logger.error("Couldn't parse ENOTTY response from an mds with error: %s\n%s" % (str(e), stdout))
+                    self.sleep_unless_stopped(5)
+            else:
+                break
+
+        return (rc, stdout)
+    
+    def get_set_state_name(self, response, set_id = None):
+        if isinstance(response, (str, bytes, bytearray)):
+            response = json.loads(response)
+
+        sets = response['sets']
+        if len(sets) == 0:
+            raise ValueError("response has no sets")
+
+        if set_id is None:
+            if len(sets) > 1:
+                raise ValueError("set_id must be provided for a multiset response")
+            else:
+                set_id = next(iter(sets.keys()))
+
+        return response['sets'][set_id]['state']['name']
+
+    def check_canceled(self, response, set_id = None):
+        if 'CANCELED' == self.get_set_state_name(response, set_id):
+            self.logger.warn('''
+                             Quiesce set got cancelled. Won't raise an error since this could be a failover, 
+                             will wait for the next quiesce attempt''')
+            return True
+        return False
+            
+    
+    def do_quiesce(self, duration):
+        
+        start_time = time.time()
+        self.logger.info(f"Going to quiesce for duration: {duration}")
+
+        # quiesce the root
+        rc, stdout = self.tell_quiesce_leader(
+            "/", # quiesce at the root
+            "--timeout", str(self.quiesce_timeout), 
+            "--expiration", str(duration + 60), # give us a minute to run the release command
+            "--await" # block until quiesced (or timedout)
+        )
+
+        try:
+            response = json.loads(stdout)
+            set_id = next(iter(response["sets"].keys()))
+        except Exception as e:
+            self.logger.error(f"Couldn't parse response with error {e}; stdout:\n{stdout}")
+            raise RuntimeError(f"Error parsing quiesce response: {e}")
+
+
+        if self.check_canceled(response):
+            return
```

Comment:
My remaining concern is that we don't know if quiesce is ever established and successfully released. Having occasional concellations is "normal" but the thrasher should fail the test if (a) it's tried a number of times over its lifetime; (b) never was successful.

Reply:
> My remaining concern is that we don't know if quiesce is ever established and successfully released

Not sure I'm following this, as we check the result of `quiesce --await` and `quiesce --release --await`.

> (a) it's tried a number of times over its lifetime; (b) never was successful.

Do you mean we should cap the times we can ignore a canceled set?

Reply:
Done